service/common/src/main/java/org/apache/polaris/service/events/listeners/FileBufferPolarisPersistenceEventListener.java

```diff
@@ -0,0 +1,189 @@
+      Clock clock) {
+    this.metaStoreManagerFactory = metaStoreManagerFactory;
+    this.polarisConfigurationStore = polarisConfigurationStore;
+    this.clock = clock;
+    shardCount =
+        polarisConfigurationStore.getConfiguration(
+            null, FeatureConfiguration.EVENT_BUFFER_NUM_SHARDS);
+    maxBufferSize =
+        polarisConfigurationStore.getConfiguration(
+            null, FeatureConfiguration.EVENT_BUFFER_MAX_SIZE);
+    int timeToFlush =
+        polarisConfigurationStore.getConfiguration(
+            null, FeatureConfiguration.EVENT_BUFFER_TIME_TO_FLUSH_IN_MS);
+    int numThreads =
+        polarisConfigurationStore.getConfiguration(
+            null, FeatureConfiguration.EVENT_BUFFER_NUM_THREADS);
+    threadPool = Executors.newScheduledThreadPool(numThreads);
+    kryo.register(PolarisEvent.class);
+    kryo.register(PolarisEvent.ResourceType.class);
+
+    // Start BufferListingTask
+    Function<FileBufferListingTask.TaskSubmissionInput, Future> taskSubmissionFunction =
+        input -> threadPool.schedule(input.task(), input.delayInMs(), TimeUnit.MILLISECONDS);
+    BiConsumer<String, List<PolarisEvent>> eventWriter =
+        (realmId, polarisEvents) -> getBasePersistenceInstance(realmId).writeEvents(polarisEvents);
+    var future =
+        threadPool.schedule(
+            new FileBufferListingTask(
+                getBufferDirectory(),
+                taskSubmissionFunction,
+                activeFlushFutures,
+                timeToFlush,
+                clock,
+                this::rotateShard,
+                eventWriter),
+            timeToFlush,
+            TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  void addToBuffer(PolarisEvent polarisEvent, CallContext callCtx) {
+    int shardNum = polarisEvent.hashCode() % shardCount;
+    String realmId = callCtx.getRealmContext().getRealmIdentifier();
+    if (!buffers.containsKey(realmId)) {
+      createBuffersForRealm(realmId);
+    }
+    BufferShard bufferShard = buffers.get(realmId).get(shardNum);
+    if (bufferShard == null) {
+      LOGGER.error(
+          "No buffer shard found for realm: #{}, shard #{}. Event dropped: {}",
+          realmId,
+          shardNum,
+          polarisEvent);
+      return;
+    }
+
+    kryo.writeObject(bufferShard.output, polarisEvent);
+    bufferShard.output.flush();
+
+    // If too many events in this buffer shard, start a new shard
+    int bufferEventCount = bufferShard.eventCount.getAndIncrement() + 1;
+    if (bufferEventCount >= maxBufferSize) {
+      rotateShard(realmId, shardNum);
+    }
+  }
+
+  private void createBuffersForRealm(String realmId) {
+    HashMap<Integer, BufferShard> bufferShardsForRealm = new HashMap<>();
+    buffers.put(realmId, bufferShardsForRealm);
+    for (int i = 0; i < shardCount; i++) {
+      bufferShardsForRealm.put(i, createShard(realmId, i));
+    }
+  }
+
+  private BufferShard createShard(String realmId, int shardNum) {
+    String bufferDirName =
```

Comment:
Instead of concatenating path segments, you should rather rely on the `Path` API.